fixes #15610 - choose firmware for vsphere vms

timogoebel opened the pull request:
I know efi support is still on it's way, but why not start with this.

We can then later either
- remove the radio buttons
- add an "auto" radio button that passed the correct parameter to fog

@lzap : Do you already have something in mind here?

timogoebel committed:
fixes #15610 - choose firmware for vsphere vms

lzap commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -142,6 +142,13 @@ def scsi_controller_types
+    def firmware_types
+      {
+        "bios" => N_("BIOS"),
+        "efi" => N_("EFI")
```
Wikipedia says that UEFI is the standard now, I've decided to go with UEFI everywhere in Foreman core for now.

timogoebel replied:
I think we should change the priority at the time there is EFI support in core.

lzap commented on app/views/compute_resources_vms/form/vmware/_base.html.erb:
```diff
@@ -3,6 +3,12 @@
+<%= field(f, :firmware, :label => _('Firmware'), :label_size => "col-md-2") do
+  compute_resource.firmware_types.collect do |type, name|
+    radio_button_f f, :firmware, {:disabled => !new_host, :value => type,
+        :text => _(name)}
```
I'd prefer not to wrap this one.

But these were all nitpicks. :+1: but I don't have an instance to play with.

timogoebel replied:
What do you mean by not wrapping these? Hardcode the options and not loop over them?

lzap replied:
> What do you mean by not wrapping these? Hardcode the options and not loop over them?

Oh no, we tend to wrap lines at 100+ column and not at 80th... no
biggie.

## 

Later,
 Lukas #lzap Zapletal

xprazak2 commented:
LGTM

timogoebel commented:
@lzap , @xprazak2 : Rebased and added an "automatic" mode that chooses the Firmware based on the pxe_loader flag of the host.

lzap commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -503,8 +513,18 @@ def vm_instance_defaults
+
+    def firmware_mapping(firmware)
+      case firmware
+      when /UEFI/
```
This should be `/(UEFI/SecureBoot)/` because we have them separate.

timogoebel replied:
I chose the vmware model because I believe it is a better place as the mapping is completely vmware specific. Do you tend to differ?

lzap replied:
I'd expect the very same for other providers like libvirt or oVirt, the method could return :uefi or :bios respectively.

timogoebel replied:
Yes, but than we'd have to map this to VMWare's "efi" or "bios". I think, this is out of scope here and can easily be extracted when this is added to other compute provisders.

lzap commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -503,8 +513,18 @@ def vm_instance_defaults
+
+    def firmware_mapping(firmware)
+      case firmware
+      when /UEFI/
```
Feel free to create a helper in the pxe_loader helper class `PxeLoaderSupport` (and a unit test) for that. That's more appropriate place than here.

lzap replied:
Scratch the SecureBoot comment, there is "UEFI" there so it's valid. But I'd suggest to move it and add a unit test.

lzap commented:
Other than that +1 but I have no VmWare to test with. @dLobatog @ohadlevy who is merging VmWare stuff at the moment?

timogoebel commented:
Rebased.

lzap commented on app/models/compute_resources/foreman/model/vmware.rb:
```diff
@@ -503,8 +513,18 @@ def vm_instance_defaults
+      when /UEFI/
+        'efi'
+      else
+        'bios'
```
One more note, looks like all cases falls down here, even non-Intel platforms. I doubt VmWare supports anything other than x86, you could make an extra check for architecture. But BIOS is a sane default I think, less code less bugs. +1

lzap reviewed (approved):
I have no instance to test with.

timogoebel commented:
@lzap: I implemented your suggestions and moved the firmware mapping to the Pxe-Helper as suggested. I still have to test this, but could you take a look at the code. I'd like to get this shipped.
Would you merge, if I sent you a video demonstrating the functionality?
